Initial docs for ENSDb (#1942)
namehash/ensnode: docs/ensnode.io/src/content/docs/ensdb/index.mdx

---
title: Getting started with ENSDb
description: ENSDb is an open standard for bi-directional ENS integration. An ENSDb instance is a PostgreSQL database served from a PostgreSQL server — and a single server can serve multiple ENSDb instances.
sidebar:
  label: Overview
  order: 1
---

import { LinkCard, Aside, Card, CardGrid } from '@astrojs/starlight/components';

## Vision

Get the whole, live onchain state of ENS in your database.

## Core Philosophy

### Open Standard

ENSDb is an open standard for bi-directional ENS integration. It defines a carefully crafted set of database schema designs, rules, and constraints for storing the ENS onchain state in a PostgreSQL database — making the data accessible from any programming language.

> The [ENSDb standard](/ensdb/concepts/glossary#ensdb-standard) is implementation-agnostic, so you have two options: implement your own [ENSDb Writer](/ensdb/concepts/glossary#ensdb-writer) and [ENSDb Reader](/ensdb/concepts/glossary#ensdb-reader) in any language, or use the reference implementations provided by ENSNode — [ENSIndexer](/ensindexer/) as the ENSDb Writer and [ENSApi](/ensapi/) as the ENSDb Reader.

Reviewer: ```suggestion
The [ENSDb standard](/ensdb/concepts/glossary#ensdb-standard) is implementation-agnostic, so you have options: implement your own [ENSDb Writer](/ensdb/concepts/glossary#ensdb-writer) or [ENSDb Reader](/ensdb/concepts/glossary#ensdb-reader) in any language, or use the initial reference implementations provided by ENSNode — [ENSIndexer](/ensindexer/) as an ENSDb Writer and [ENSApi](/ensapi/) as an ENSDb Reader.
```

<Aside type="tip" title="Build in Any Language">

> Each ENSDb instance is a standard PostgreSQL database, so to interact with it, you can use _any programming language_ that has a PostgreSQL driver. Zig, Rust, Go, Python, JavaScript, Ruby, Java, C# — the choice is yours.

Reviewer: ```suggestion
Each ENSDb instance runs in a standard PostgreSQL database, so to interact with it, you can use _any programming language_ that has a PostgreSQL driver. Zig, Rust, Go, Python, JavaScript, Ruby, Java, C# — the choice is yours.
```

</Aside>

## What You Get

### Complete ENS State

> An **ENSDb instance** contains the **onchain state of ENS**, including:

Reviewer: ```suggestion
An **ENSDb instance** contains the **live onchain state of ENS**, including:
```

- Ecosystem — Mature tools, ORMs, dashboards, analytics platforms
- Reliability — Decades of production-proven technology

## What You Can Build

ENSDb unlocks a new universe of ENS applications:

<CardGrid>
<Card title="Custom APIs" icon="seti:graphql">
Build specialized GraphQL or REST APIs tailored to your use case. Query exactly the data you need with full SQL power.
</Card>
<Card title="Analytics & Dashboards" icon="document">
Create real-time dashboards and analytics pipelines. Better than Dune — you have the live ENS state locally with sub-second query latency.
</Card>
<Card title="CLIs & Developer Tools" icon="seti:shell">
Build command-line tools for ENS operations. Query domains, and any specialized lookup — all from your terminal.
</Card>
<Card title="Event-Based Engines" icon="seti:plan">
Build reactive systems that respond to ENS state changes. Respond to registration lifecycle updates, ownership transfers, resolver updates.
</Card>
<Card title="Data Pipelines" icon="seti:plan">
Feed ENS data into your existing data infrastructure. Sync to data warehouses, trigger webhooks, populate search indexes.
</Card>
<Card title="AI Agents" icon="seti:bicep">

> Build AI agents that can query the ENS state and perform actions based on that data.

Reviewer: ```suggestion
Build AI agents that can query ENS state and perform actions based on that data.
```
